File: provisa-ui/src/lib/orgBranding.ts
// Copyright (c) 2026 Kenneth Stott
// Canary: a9f889bb-8065-40b7-96a5-bd05c1f05a63
//
// This source code is licensed under the Business Source License 1.1
// found in the LICENSE file in the root directory of this source tree.
//
// NOTICE: Use of this software for training artificial intelligence or
// machine learning models is strictly prohibited without explicit written
// permission from the copyright holder.

// REQ-1486: apply an org's branding to the running page.
//
// Branding is light by design: it moves the accent colors and names the org, it does not re-skin
// the product. Applying it means overriding a handful of theme tokens on :root — the same
// variables theme/tokens.css defines — so every component that already reads them follows without
// knowing an org set anything. Light and dark are both driven by these tokens, so the override
// holds across a theme switch.

import type { OrgBranding } from "../api/branding";
import { isOrgSubdomainHost, orgFromHost } from "./authHost";

/**
 * The tokens an org may move. primary_color drives the three existing --primary tokens, so every
 * button and active-nav state follows it. accent_color has no existing token to take over — it is
 * the org's own second color and lands on --org-accent, which the branded header reads.
 */
const PRIMARY_TOKENS = ["--primary", "--primary-hover", "--primary-strong"];
const ACCENT_TOKEN = "--org-accent";

export function applyOrgBranding(branding: OrgBranding): void {
  const root = document.documentElement;
  if (branding.primary_color) {
    for (const token of PRIMARY_TOKENS) root.style.setProperty(token, branding.primary_color);
  }
  if (branding.accent_color) root.style.setProperty(ACCENT_TOKEN, branding.accent_color);
}

/** Undo an applied override, so signing out of one org does not leave its colors behind. */
export function clearOrgBranding(): void {
  const root = document.documentElement;
  for (const token of [...PRIMARY_TOKENS, ACCENT_TOKEN]) root.style.removeProperty(token);
}

/**
 * The org whose branding this page should show, or null when the page belongs to no org.
 *
 * On an org subdomain that is the Host's own org. On the control plane it is the org named by
 * `?next=` — the address the user was sent here from and will be returned to (REQ-1348) — which is
 * the only org this sign-in is about. A control-plane sign-in reached directly names none.
 */
export function brandingOrg(search: string = window.location.search): string | null {
  if (isOrgSubdomainHost()) return orgFromHost();
  const next = new URLSearchParams(search).get("next");
  if (!next) return null;
  try {
    const host = new URL(next).hostname;
    return isOrgSubdomainHost(host) ? orgFromHost(host) : null;
  } catch {
    return null;
  }
}

File: provisa/core/org_branding.py
# Copyright (c) 2026 Kenneth Stott
#
# This source code is licensed under the Business Source License 1.1
# found in the LICENSE file in the root directory of this source tree.
#
# NOTICE: Use of this software for training artificial intelligence or
# machine learning models is strictly prohibited without explicit written
# permission from the copyright holder.

"""Per-org branding (REQ-1486).

An org_admin may put their own mark on the surfaces their people meet before and just after
sign-in: a display name, a logo, two colors, a welcome line, and the sentence the invitation adds
to Provisa's own copy. It is deliberately light — a tenant styles its instance, it does not
re-skin the product — so the field set is fixed here and every value is validated before it is
stored, not when it is rendered.

Storage is one JSON document on the org row plus the logo bytes beside it (the logo is served
from the control plane, so nothing on the page depends on a tenant-supplied remote URL).

Validation is strict and total: a value that does not parse is refused at the edit, so every
reader can render what it reads. There is no defaulting here — an unset field is absent from the
document, and what an unset field looks like is the UI's own design, not a substituted value.
"""

from __future__ import annotations

import json
import re
from typing import Any

# The editable field set. Fixed: branding is a tenant mark on Provisa's UI, not a theme engine.
FIELDS = (
    "display_name",
    "primary_color",
    "accent_color",
    "welcome_message",
    "invite_message",
)

_HEX_COLOR = re.compile(r"^#[0-9a-fA-F]{6}$")

# Caps sized to the surface each value lands on: a header slot, a login sub-heading, a paragraph
# in an email. A longer value would not be rejected by the renderer, it would just break it.
_MAX_LEN = {
    "display_name": 80,
    "welcome_message": 400,
    "invite_message": 1000,
}

# Logo formats every browser and mail client renders, and a size that stays inline in a JSON-free
# byte response. SVG is accepted because a wordmark is usually vector; it is served with a
# Content-Security-Policy that keeps it from executing (see the branding router).
LOGO_MEDIA_TYPES = ("image/png", "image/jpeg", "image/svg+xml", "image/webp")
MAX_LOGO_BYTES = 256 * 1024


class BrandingError(ValueError):
    """A branding value that cannot be stored. Carries the field so the API can name it."""

    def __init__(self, field: str, message: str) -> None:
        super().__init__(message)
        self.field = field


def validate_branding(payload: dict[str, Any]) -> dict[str, str]:
    """The branding document as it will be stored, or raise.

    Unknown keys are refused rather than dropped: an org_admin who misspells a field must be told,
    because silently storing nothing looks identical to storing a value that has no effect.
    An explicit ``None`` clears the field, which is how the UI's "reset" works.
    """
    unknown = sorted(set(payload) - set(FIELDS))
    if unknown:
        raise BrandingError(unknown[0], f"Unknown branding field(s): {', '.join(unknown)}")
    document: dict[str, str] = {}
    for field in FIELDS:
        value = payload.get(field)
        if value is None:
            continue
        if not isinstance(value, str):
            raise BrandingError(field, f"{field} must be a string")
        value = value.strip()
        if not value:
            continue  # an emptied input clears the field, same as omitting it
        if field.endswith("_color"):
            if not _HEX_COLOR.match(value):
                raise BrandingError(field, f"{field} must be a hex color like #4f46e5")
            document[field] = value.lower()
            continue
        limit = _MAX_LEN[field]
        if len(value) > limit:
            raise BrandingError(field, f"{field} must be {limit} characters or fewer")
        document[field] = value
    return document


def serialize_branding(document: dict[str, str]) -> str | None:
    """The column value for a validated document. An empty document stores NULL — "this org has
    set no branding" is the absence of a row value, not an empty object to be parsed."""
    return json.dumps(document, sort_keys=True) if document else None


def parse_branding(column_value: str | None) -> dict[str, str]:
    """The document a stored column holds. NULL means no branding was ever set.

    A non-NULL value that does not parse is a corrupted row, not a missing one, so it raises:
    every value in this column was written through :func:`validate_branding`.
    """
    if column_value is None:
        return {}
    document = json.loads(column_value)
    if not isinstance(document, dict):
        raise BrandingError("branding", "Stored branding is not an object")
    return document


def validate_logo(data: bytes, media_type: str) -> str:
    """The normalized media type for an accepted logo, or raise."""
    normalized = media_type.split(";")[0].strip().lower()
    if normalized not in LOGO_MEDIA_TYPES:
        raise BrandingError(
            "logo",
            f"Logo must be one of: {', '.join(LOGO_MEDIA_TYPES)} (got {normalized or 'nothing'})",
        )
    if not data:
        raise BrandingError("logo", "Logo is empty")
    if len(data) > MAX_LOGO_BYTES:
        raise BrandingError(
            "logo", f"Logo must be {MAX_LOGO_BYTES // 1024} KiB or smaller ({len(data)} bytes)"
        )
    return normalized
